selfdrive/car/body/carcontroller.py

```diff
@@ -66,25 +87,68 @@ def update(self, CC, CS):
                                        self.torque_l_filtered - MAX_TORQUE_RATE,
                                        self.torque_l_filtered + MAX_TORQUE_RATE)
-      torque_r = int(np.clip(self.torque_r_filtered, -MAX_TORQUE, MAX_TORQUE))
+      self.torque_r_filtered = np.clip(self.deadband_filter(torque_r, 10),
+                                       self.torque_r_filtered - MAX_TORQUE_RATE,
+                                       self.torque_r_filtered + MAX_TORQUE_RATE)
+
       torque_l = int(np.clip(self.torque_l_filtered, -MAX_TORQUE, MAX_TORQUE))
+      torque_r = int(np.clip(self.torque_r_filtered, -MAX_TORQUE, MAX_TORQUE))
+
+    # Knee angle controls
+    if self.CP.carFingerprint == CAR.BODY_KNEE and CC.enabled:
+      knee_angle_desired_left = CC.actuators.gas
+      knee_angle_desired_right = CC.actuators.brake
+
+      knee_angle_measured_left = CS.knee_angle_l
+      knee_angle_measured_right = CS.knee_angle_r
+
+      knee_angle_error_left = knee_angle_desired_left - knee_angle_measured_left
+      knee_angle_error_right = knee_angle_desired_right - knee_angle_measured_right
+
+      freeze_integrator = ((knee_angle_error_left < 0 and self.knee_angle_left_pid.error_integral <= -MAX_KNEE_ANGLE_INTEGRATOR_LEFT) or
+                           (knee_angle_error_left > 0 and self.knee_angle_left_pid.error_integral >= MAX_KNEE_ANGLE_INTEGRATOR_LEFT))
+      knee_torque_l = self.knee_angle_left_pid.update(knee_angle_error_left, freeze_integrator=freeze_integrator)
+
+      freeze_integrator = ((knee_angle_error_right < 0 and self.knee_angle_right_pid.error_integral <= -MAX_KNEE_ANGLE_INTEGRATOR_RIGHT) or
+                           (knee_angle_error_right > 0 and self.knee_angle_right_pid.error_integral >= MAX_KNEE_ANGLE_INTEGRATOR_RIGHT))
+      knee_torque_r = self.knee_angle_right_pid.update(knee_angle_error_right, freeze_integrator=freeze_integrator)
+
+    # Torque rate limits
+      self.torque_knee_l_filtered = np.clip(knee_torque_l,
+                                        self.torque_knee_l_filtered - MAX_KNEE_TORQUE_RATE,
+                                        self.torque_knee_l_filtered + MAX_KNEE_TORQUE_RATE)
+      self.torque_knee_r_filtered = np.clip(knee_torque_r,
+                                        self.torque_knee_r_filtered - MAX_KNEE_TORQUE_RATE,
+                                        self.torque_knee_r_filtered + MAX_KNEE_TORQUE_RATE)
+
+      knee_torque_l = int(np.clip(self.torque_knee_l_filtered, -MAX_KNEE_TORQUE_LEFT, MAX_KNEE_TORQUE_LEFT))
+      knee_torque_r = int(np.clip(self.torque_knee_r_filtered, -MAX_KNEE_TORQUE_RIGHT, MAX_KNEE_TORQUE_RIGHT))
+
+      # Extra safety for when training stands attached:
+      if knee_angle_measured_left<=30 and knee_torque_l > 0:
+        knee_torque_l = 0
+      elif knee_angle_measured_left >= 330 and knee_torque_l < 0:
+        knee_torque_l = 0
+
+      # Do not try to balance if knee and hip have excessive angle on start
```

Comment:
is the rate limit low enough to safely fade in control if a body is already onroad and the user moves it into position?

Reply:
What do you mean by "moves it into position" ? moves body or knee?